pkg/api/errors.go

```diff
@@ -0,0 +1,173 @@
+package api
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+)
+
+const (
+	contentType = "Content-Type"
+)
+
+var jsonTypeRE = regexp.MustCompile(`[/+]json($|;)`)
+
+// HTTPError represents an error response from the GitHub API.
+type HTTPError struct {
+	AcceptedOAuthScopes string
+	Errors              []HTTPErrorItem
+	Message             string
+	OAuthScopes         string
+	RequestURL          *url.URL
+	StatusCode          int
+}
+
+// HTTPErrorItem stores additional information about an error response
+// returned from the GitHub API.
+type HTTPErrorItem struct {
+	Message  string
+	Resource string
+	Field    string
+	Code     string
+}
+
+// Allow HTTPError to satisfy error interface.
+func (err HTTPError) Error() string {
+	if msgs := strings.SplitN(err.Message, "\n", 2); len(msgs) > 1 {
+		return fmt.Sprintf("HTTP %d: %s (%s)\n%s", err.StatusCode, msgs[0], err.RequestURL, msgs[1])
+	} else if err.Message != "" {
+		return fmt.Sprintf("HTTP %d: %s (%s)", err.StatusCode, err.Message, err.RequestURL)
+	}
+	return fmt.Sprintf("HTTP %d (%s)", err.StatusCode, err.RequestURL)
+}
+
+// GQLError represents an error response from GitHub GraphQL API.
+type GQLError struct {
+	Errors []GQLErrorItem
+}
+
+// GQLErrorItem stores additional information about an error response
+// returned from the GitHub GraphQL API.
+type GQLErrorItem struct {
+	Message string
+	Path    []interface{}
+	Type    string
+}
+
+// Allow GQLError to satisfy error interface.
+func (gr GQLError) Error() string {
+	errorMessages := make([]string, 0, len(gr.Errors))
+	for _, e := range gr.Errors {
+		msg := e.Message
+		if p := e.pathString(); p != "" {
+			msg = fmt.Sprintf("%s (%s)", msg, p)
+		}
+		errorMessages = append(errorMessages, msg)
+	}
+	return fmt.Sprintf("GQL: %s", strings.Join(errorMessages, ", "))
+}
+
+// Match determines if the GQLError is about a specific type on a specific path.
+// If the path argument ends with a ".", it will match all its subpaths.
+func (gr GQLError) Match(expectType, expectPath string) bool {
+	for _, e := range gr.Errors {
+		if e.Type != expectType || !matchPath(e.pathString(), expectPath) {
+			return false
+		}
+	}
+	return true
+}
+
+func (ge GQLErrorItem) pathString() string {
+	var res strings.Builder
+	for i, v := range ge.Path {
+		if i > 0 {
+			res.WriteRune('.')
+		}
+		fmt.Fprintf(&res, "%v", v)
+	}
+	return res.String()
+}
+
+func matchPath(p, expect string) bool {
+	if strings.HasSuffix(expect, ".") {
+		return strings.HasPrefix(p, expect) || p == strings.TrimSuffix(expect, ".")
+	}
+	return p == expect
+}
+
+// HandleHTTPError parses a http.Response into a HTTPError.
+func HandleHTTPError(resp *http.Response) error {
```

Comment:
I am not in love with this name, I am open to suggestions on a better one.

Reply:
I think it's fine! Ideally, no integrators will ever need to call this directly, since all the other higher-level API methods call this internally on non-HTTP 200 statuses. However, I understand we need this from `cli/cli`?

Reply:
Yeah, there are numerous places ing `cli/cli` that call `api.HandleHTTPError` directly rather than making requests through an `api.Client` interface.